### DISCUSSION TITLE
posts/handle-loop-of-callback-in-javascript

### DISCUSSION THREAD

**Discussion by giscus** [mildronize/blog-v8#46] [Blog Comment]
# posts/handle-loop-of-callback-in-javascript

My personal blog

https://www.thadaw.com/posts/handle-loop-of-callback-in-javascript?id=jwof8mf&fbclid=IwZXh0bgNhZW0CMTEAAR2Iu_-iRggDIr84TC3j62kvO6IP8SxZhsNBKIsVqt1YWU-PnrKIYwXTTOI_aem_7VmY482TQK8uXvVqQncE4Q

**Comment by kasamachenkaow** [comment]
จริง node มี Util ชื่อ promisify() เอาไว้ช่วยแปลงพวก callback function อย่าง setTimeout() ให้เป็น async ด้วยนะครับ ส่วน bun จะมี sleep มาให้เลยครับ

**Reply by mildronize** [reply]
จริงด้วย เพิ่งนึกๆๆ ออก ขอบคุณนะครับ แต่ถ้าเจอปัญหาแบบนี้ ไม่สามารถใช้ `promisify()` ได้นะ
แล้วอีกอย่างผมไม่ค่อยเขียน Code ไม่ให้ขึ้นต่อ Runtime ครับ เท่าไหร่ครับ เพราะอยากให้แชร์ได้ทั้ง Broswer/Node/Bun เลย

**Reply by kasamachenkaow** [reply]
ตัวปัญหาหลักที่คุณมายด์แก้ผมว่าโอเคครับ ผมเสริมตรง delay function เท่านั้นครับ (ผมพิมพ์ไม่เคลียร์ 😁)
แต่ point เรื่อง code ที่ไม่ขึ้นกับ runtime ก็ถือว่า good point ครับ จะได้ไม่ต้องไปใช้พวก polyfill ให้วุ่นวาย

**Reply by mildronize** [reply]
อ่อๆ ใช่ๆ ตรง delay function เราสามารถใช้ `promisify()` ได้ครับ, จากประสบการณ์เท่าที่ลองคือ ปัญหาของเจ้าตัว `promisify()` คือ
1. ขึ้นกับ Node.js Runtime
2. บาง Callback ไม่สามารถใช้ `promisify()` ได้เลยครับ

ดังนั้นถ้าเรารู้จักการว่า ถ้าเราทำแปลงเป็น Promise ได้เอง จะทำให้เราแก้ปัญหาที่ว่ามาได้ง่ายขึ้นครับ 😁

**Comment by mildronize** [comment]
**แก้ไข:**

ผมเพิ่งเห็นว่าเจ้า method `index.export` มันเป็น Async อยู่แล้วมัน Return เป็น Promise 🤣

```ts

await index.export(
  (key, data) => {
    const targetIndex = path.join(searchIndexPath, `${key}.json`);
    fs.writeJSONSync(targetIndex, data ?? {});
    targetIndexPath.push(targetIndex);
  }
);
```

**Reply by kasamachenkaow** [reply]
โอ้ 55 ง่ายขึ้นเยอะครับ

**Reply by mildronize** [reply]
เขินเบยยย แต่ก็นะเผื่อเจอเคสนี้จริงๆ ก็ได้เอาความรู้ไปประยุกต์ใช้ละน้าาา ฮ่าๆๆๆ
